src/typechecker.erl

```diff
@@ -4339,12 +4422,16 @@ add_type_pat(OpPat = {op, _Anno, _Op, _Pat1, _Pat2}, Ty, Env, VEnv) ->
     %% the pattern has the correct type) but we also need to check that the left
     %% argument is not smaller than the right argument. But instead of implementing
     %% such a check, we simply evaluate the pattern as an expression.
-    add_type_pat_literal(OpPat, Ty, Env, VEnv);
-add_type_pat(OpPat = {op, _Anno, _Op, _Pat}, Ty, Env, VEnv) ->
-    add_type_pat_literal(OpPat, Ty, Env, VEnv);
-add_type_pat(Pat, Ty, _Env, _VEnv) ->
+    add_type_pat_literal(OpPat, Ty, Env);
+add_type_pat(OpPat = {op, _Anno, _Op, _Pat}, Ty, Env) ->
+    add_type_pat_literal(OpPat, Ty, Env);
+add_type_pat(Pat, Ty, _Env) ->
     throw({type_error, pattern, element(2, Pat), Pat, Ty}).
 
+is_list_pat_exhaustive({nil, _}) -> true;
+is_list_pat_exhaustive({cons, _, {var, _, _}, {var, _, _}}) -> true;
```

Comment:
`[X|Y]` only exhausts a nonempty list type if X and Y are unbound and distinct. Need to check that. Or does this function just check that a pattern doesn't disable exhaustiveness checking?

Reply:
> Or does this function just check that a pattern doesn't disable exhaustiveness checking?

Exactly, we use this to disable exhaustiveness checking for the current clauses list. We could definitely make it more sophisticated, but with the current tests this is enough.

Reply:
OK, so in the following example, Gradualizer thinks that the case clauses are exhaustive and doesn't return an error, but it could as well have disabled exhaustiveness checking. For the exhaustiveness checking, there's no difference, but for refinement there might be a difference. Maybe we get a warning for unreachable clauses?

```Erlang
-spec f([atom()]) -> ok.
f(X) ->
    case X of
        [] -> ok;
        [X|_] -> ok     % <--- Not exhaustive, but Gradualizer thinks it is.
        [hello|_] -> ok % <--- False positive: clause not reachable?
    end.
```

Reply:
This function is not actually checking if the list pattern is exhaustive. We can leave it as is for now, but then at least rename the function or add a comment about it. It can be hard to understand this in the future otherwise.

I think it's only called with CONS patterns, so the first clause is not used AFAICT.

Reply:
This is handled correctly in each case. Consider the following examples.

Variable X is bound and we try to match on it as the first element of itself-the-list:

```
17:09:25 erszcz @ x6 : ~/work/erszcz/gradualizer (fix-list-exhaustiveness-checking-regressions-3 %)
$ cat t.erl
-module(t).

-compile(export_all).

-spec f([atom()]) -> ok.
f(X) ->
    case X of
        [] -> ok;
        [X|_] -> ok;     % <--- Not exhaustive, but Gradualizer thinks it is.
        [hello|_] -> ok  % <--- False positive: clause not reachable?
    end.
17:09:27 erszcz @ x6 : ~/work/erszcz/gradualizer (fix-list-exhaustiveness-checking-regressions-3 %)
$ ./bin/gradualizer t.erl
t.erl: The variable on line 9 at column 10 is expected to have type atom() but it has type [atom()]

    case X of
        [] -> ok;
        [X|_] -> ok;     % <--- Not exhaustive, but Gradualizer thinks it is.
         ^
```

A true positive error is reported - `X is expected to have type atom() but it has type [atom()]`.

We use a free variable Y to match on the first element of the list X:

```
17:09:28 erszcz @ x6 : ~/work/erszcz/gradualizer (fix-list-exhaustiveness-checking-regressions-3 %)
$ cat t.erl
-module(t).

-compile(export_all).

-spec f([atom()]) -> ok.
f(X) ->
    case X of
        [] -> ok;
        [Y|_] -> ok;     % <--- Not exhaustive, but Gradualizer thinks it is.
        [hello|_] -> ok  % <--- False positive: clause not reachable?
    end.
17:09:33 erszcz @ x6 : ~/work/erszcz/gradualizer (fix-list-exhaustiveness-checking-regressions-3 %)
$ ./bin/gradualizer t.erl
t.erl: The clause on line 10 at column 9 cannot be reached
```

A true positive error is reported - since we match on anything with Y, the following clause cannot be reached.

What if we invert the order of Y and `hello`?

```
17:09:35 erszcz @ x6 : ~/work/erszcz/gradualizer (fix-list-exhaustiveness-checking-regressions-3 %)
$ cat t.erl
-module(t).

-compile(export_all).

-spec f([atom()]) -> ok.
f(X) ->
    case X of
        [] -> ok;
        [hello|_] -> ok;  % <--- False positive: clause not reachable?
        [Y|_] -> ok     % <--- Not exhaustive, but Gradualizer thinks it is.
    end.
17:09:52 erszcz @ x6 : ~/work/erszcz/gradualizer (fix-list-exhaustiveness-checking-regressions-3 %)
$ ./bin/gradualizer t.erl
17:09:55 erszcz @ x6 : ~/work/erszcz/gradualizer (fix-list-exhaustiveness-checking-regressions-3 %)
$
```

Nothing is reported as it's valid code - we have a specific clause matching on `hello` and a following catch-all with Y.

All good :D